numba/parfor.py

```diff
@@ -1101,12 +1384,66 @@ def _reduce_to_parfor(self, equiv_set, lhs, expr):
                                         self.typemap,
                                         self.calltypes)
         loop_body = reduce_f_ir.blocks
-        acc_label = next_label()
-        loop_body[acc_label] = acc_block
-        first_reduce_block = reduce_f_ir.blocks[min(reduce_f_ir.blocks.keys())]
-        first_reduce_block.body.insert(0, ir.Assign(getitem_call, tmp_var, loc))
+        end_label = next_label()
+        end_block = ir.Block(scope, loc)
+        loop_body[end_label] = end_block
+        first_reduce_label = min(reduce_f_ir.blocks.keys())
+        first_reduce_block = reduce_f_ir.blocks[first_reduce_label]
+        body_block.body.extend(first_reduce_block.body)
+        first_reduce_block.body = body_block.body
         replace_arg_nodes(first_reduce_block, [acc_var, tmp_var])
-        replace_returns(loop_body, acc_var, acc_label)
+        replace_returns(loop_body, acc_var, end_label)
+        return index_var, loop_body
+
+    def _reduce_to_parfor(self, equiv_set, lhs, args, loc):
+        """
+        Convert a reduce call to a parfor.
+        The call arguments should be (call_name, array, init_value).
+        """
+        scope = lhs.scope
+        call_name = args[0]
+        in_arr = args[1]
+        arr_def = get_definition(self.func_ir, in_arr.name)
+
+        mask_var = None
+        mask_indices = None
+        result = guard(self._find_mask, arr_def)
+        if result:
+            in_arr, mask_var, mask_typ, mask_indices = result
```

Comment:
Am I right in reading this as reduction-with-mask support?

Reply:
Yes, `_reduce_to_parfor` handles `functools.reduce` case specifically.